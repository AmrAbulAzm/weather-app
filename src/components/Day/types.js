import { shape, string, number, bool, array, arrayOf, func } from 'prop-types'

const weatherItem = shape({
  dt: number,
  main: shape({
    temp: number.isRequired,
    temp_min: number,
    temp_max: number,
    pressure: number,
    sea_level: number,
    grnd_level: number,
    humidity: number,
    temp_kf: number
  }),
  weather: arrayOf(shape({
    id: number,
    main: string.isRequired,
    description: string,
    icon: string,
  })),
  clouds: shape({
    all: number
  }),
  wind: shape({
    speed: number,
    deg: number
  }),
  rain: shape({
    ['3h']: number
  }),
  sys: shape({
    pod: string
  }),
  dt_txt: string.isRequired
})

const DayProptypes = {
  data: arrayOf(weatherItem).isRequired,
  city: string.isRequired,
  maxMin: shape({
    max: number,
    min: number
  }).isRequired,
  weekDay: string.isRequired,
  day: string.isRequired,
  month: string.isRequired
}

const WeatherProptypes = {
  data: arrayOf(weatherItem).isRequired,
  current: weatherItem.isRequired,
  setCurrent: func.isRequired
}

const WeatherItemProptypes = {
  dayTime: string.isRequired,
  condition: string.isRequired,
  temperature: number.isRequired,
  selected: bool.isRequired,
  onClick: func.isRequired
}

export {
  DayProptypes,
  WeatherProptypes,
  WeatherItemProptypes
}