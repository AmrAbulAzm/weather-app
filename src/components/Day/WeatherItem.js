import React from 'react';
import styled, { css } from 'styled-components';
import { Icon } from '../../utils/helpers';
import { WeatherItemProptypes } from './types';

const WeatherItem = ({
  dayTime,
  condition,
  temperature,
  selected,
  onClick
}) => (
  <WeatherItemButton selected={selected} onClick={onClick}>
    <SubText>{dayTime}</SubText>
    <Image src={Icon[condition]} alt={'smallWeatherIcon'} /> 
    <SmallTemperature>{`${temperature}Â°`}</SmallTemperature>
  </WeatherItemButton>
);

WeatherItem.propTypes = WeatherItemProptypes;

const WeatherItemButton = styled.button`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 22px;
  border: none;
  background: none;
  outline: none;

  ${props => props.selected && css`
    border-radius: 6px;
    background-color: #51557A;
  `}
`;

const Image = styled.img`
  width: 80px;
  margin: 10px 0px;
`;

const SubText = styled.p`
  font-size: 32px;
  color: #A8AABD;
  margin: 0px;
`;

const SmallTemperature = styled.h3`
  font-size: 55px;
  color: white;
  margin: 0px;
`;

export default WeatherItem;