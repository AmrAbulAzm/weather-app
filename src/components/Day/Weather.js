import React from 'react';
import styled from 'styled-components';
import WeatherItem from './WeatherItem';
import { WeatherProptypes } from './types';

const Weather = ({ data, current, setCurrent }) => (
  <Wrap>
    {data.map(el => {
      const temperature = Math.round(el.main.temp - 273.15)
      const condition = el.weather[0].main
      const time = el.dt_txt.split(' ')[1]
      const formattedTime = time.substring(0, time.length - 3)
      const selected = el.dt_txt === current.dt_txt ? true : false;
      return (
        <WeatherItem
          key={el.dt_txt}
          dayTime={formattedTime}
          condition={condition}
          temperature={temperature}
          selected={selected}
          onClick={() => setCurrent(el)}
        />
      )
    })}
  </Wrap>
)

Weather.propTypes = WeatherProptypes;

const Wrap = styled.main`
  display: flex;
  width: 100%;
  justify-content: space-between;
  margin-top: 40px;
  padding: 10px;
  box-sizing: border-box;
  overflow-x: scroll;
`;

export default Weather;