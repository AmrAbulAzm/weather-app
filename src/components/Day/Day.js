import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import Weather from './Weather'
import { Icon } from '../../utils/helpers';

const Day = ({ data, city, maxMin, weekDay, day, month }) => {
  const [current, setCurrent] = useState(data[0])

  useEffect(() => setCurrent(data[0]), [data])
  return (
    <Wrap>
      <header>
        
        <section>
          <img src={Icon[current.weather[0].main]} style={{ width: 220 }} /> 
        </section>

        <section>

          <ValuesContainer>
            <SubText>{current.weather[0].main}</SubText>
            <SubText>{`${maxMin.max}° / ${maxMin.min}°`}</SubText>
          </ValuesContainer>

          <Temperature>{`${Math.round(current.main.temp  - 273.15)}°`}</Temperature>

        </section>


        <LeftAlignedSection>

          <SubText style={{ marginBottom: 22 }}>{city}</SubText>

          <div>
            <DateInfo>{weekDay}</DateInfo>
            <DateInfo>{`${day}. ${month}`}</DateInfo>
          </div>
        </LeftAlignedSection>
        

      </header>

      <Weather data={data} current={current} setCurrent={setCurrent} />
    </Wrap>
  )
}

const Wrap = styled.div`
  width: 100%;

  > header {
    display: flex;
    align-items: center;
    justify-content: space-evenly;
    margin-top: 40px;

    @media (max-width: 768px) {
      flex-direction: column;
      
      > section {
        width: 90%;
      }
    }
  }
`;

const ValuesContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin-bottom: 22px;
  width: 298px;

  @media (max-width: 768px) {
    width: 100%;
  }
`

const LeftAlignedSection = styled.section`
  text-align: left;
`


const SubText = styled.p`
  font-size: 32px;
  color: #A8AABD;
  margin: 0;
`

const Temperature = styled.h1`
  font-size: 165px;
  color: white;
  margin: 0;
`

const DateInfo = styled.h2`
  font-size: 66px;
  color: white;
  margin: 15px 0px;
`

export default Day;