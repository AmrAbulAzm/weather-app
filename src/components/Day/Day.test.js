import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Day from '.';

it('Day renders without crashing', () => {
  const props = {
    data: [{
      main: { temp: 123 },
      weather: [{ main: 'sunny' }],
      dt_txt: '2017-02-17 15:00:00'
    }],
    city: 'Berlin',
    maxMin: { max: 321, min: 123 },
    weekDay: 'Friday',
    day: '13',
    month: 'October'
  }
  const { asFragment } = render(<Day {...props} />);
  expect(asFragment()).toMatchSnapshot();
});