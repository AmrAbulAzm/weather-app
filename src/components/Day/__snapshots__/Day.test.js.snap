// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-AxgMl ewNxMU"
  >
    <header>
      <section>
        <img
          alt="bigWeatherIcon"
          class="sc-Axmtr bhtEpz"
        />
      </section>
      <section>
        <div
          class="sc-AxheI ikEUcU"
        >
          <p
            class="sc-fzozJi jPUxRQ"
          >
            sunny
          </p>
          <p
            class="sc-fzozJi jPUxRQ"
          >
            321째 / 123째
          </p>
        </div>
        <h1
          class="sc-fzpans hEsCVz"
        >
          -150째
        </h1>
      </section>
      <section
        class="sc-AxmLO fJWbmy"
      >
        <p
          class="sc-fzozJi sc-fzoLsD dxqxGG"
        >
          Berlin
        </p>
        <div>
          <h2
            class="sc-fzplWN iHlsnq"
          >
            Friday
          </h2>
          <h2
            class="sc-fzplWN iHlsnq"
          >
            13. October
          </h2>
        </div>
      </section>
    </header>
    <main
      class="sc-AxhUy cnZzqw"
    >
      <button
        class="sc-AxjAm cPOcFJ"
      >
        <p
          class="sc-AxiKw jKCCZu"
        >
          15:00
        </p>
        <img
          alt="smallWeatherIcon"
          class="sc-AxirZ fcqyWt"
        />
        <h3
          class="sc-AxhCb eKxkdA"
        >
          -150째
        </h3>
      </button>
    </main>
  </div>
</DocumentFragment>
`;
